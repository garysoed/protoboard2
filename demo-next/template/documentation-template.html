<style>
  #content::slotted(*) {
    padding-left: 1rem;
    padding-right: 1rem;
  }

  #title {
    padding-left: 1rem;
    padding-right: 1rem;
  }
</style>

<section flex>
  <h3 id="title" mk-display></h3>
  <slot id="content" name="content"></slot>
</section>
