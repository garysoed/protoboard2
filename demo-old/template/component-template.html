<style>
  #display {
    height: 16rem;
  }
</style>

<pbd-doc-template id="template">
  <section slot="content">
    <slot id="display" name="display"></slot>
    <hr>
    <slot name="customize"></slot>
  </section>
  <section slot="details">
    <slot name="details"></slot>
  </section>
</pbd-doc-template>
