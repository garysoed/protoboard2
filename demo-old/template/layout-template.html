<pbd-doc-template id="template">
  <section slot="content">
    <slot name="customize"></slot>
    <mk-text-icon-button
        icon="add"
        id="setLayout"
        label="Set Layout"
    ></mk-text-icon-button>
  </section>
  <section slot="details">
    <slot name="details"></slot>
  </section>
</pbd-doc-template>
