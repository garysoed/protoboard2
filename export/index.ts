export {Lens, $lens} from '../src/core/lens';
export {LensDisplay} from '../src/core/lens-display';

// action
export {pickAction} from '../src/action/pick-action';

// coordinate
export {Cartesian} from '../src/coordinate/cartesian';
export {Coordinate} from '../src/coordinate/coordinate';
export {Indexed} from '../src/coordinate/indexed';

// region
export {Deck, $deck, DeckSpec} from '../src/region/deck';
export {Slot, $slot, SlotSpec, slotSpec} from '../src/region/slot';

// core
export {$createSpecEntries, $createSpecMap} from '../src/objects/object-create-spec';
export {ContentSpec} from '../src/payload/is-container';
export {ObjectSpec} from '../src/types/object-spec';
export {$render, Render} from '../src/objects/render';
export {$active, Active, activeSpec} from '../src/core/active';
export {$$activeSpec} from '../src/objects/active-spec';

// piece
export {$d1, D1, D1Spec, d1Spec} from '../src/piece/d1';
export {$d2, D2, D2Spec} from '../src/piece/d2';
export {$d6, D6, D6Spec} from '../src/piece/d6';

// tiling
export {rectOrthogonal, RectOrthogonal, Direction as RectOrthogonalDirection, distance as rectOrthogonalDistance} from '../src/tiling/rect-orthogonal';
export {vhex, VHex, Direction as VHexDirection, distance as vhexDistance} from '../src/tiling/vhex';
export {Tile} from '../src/tiling/types';
