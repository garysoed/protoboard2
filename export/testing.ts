export {LensHarness} from '../src/face/testing/lens-harness';
export {createRenderSpec, TEST_FACE} from '../src/testing/test-face';