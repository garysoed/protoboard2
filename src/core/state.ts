import { Observable } from 'rxjs';

/**
 * State of a game during its runtime.
 *
 * @thModule core
 */
export interface State {
  /**
   * Identifies the object. Must be globally unique.
   */
  readonly id: string;

  /**
   * Used to map to a function that generates the object.
   */
  readonly type: string;

  /**
   * Used to generate the object.
   */
  readonly payload: ReadonlyMap<string, Observable<unknown>>;
}
