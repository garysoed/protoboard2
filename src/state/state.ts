import { BehaviorSubject } from 'rxjs';

/**
 * State of a game during its runtime.
 *
 * @thModule core
 */
export interface State<P extends object> {
  /**
   * Identifies the object. Must be globally unique.
   */
  readonly id: string;

  /**
   * Used to map to a function that generates the object.
   */
  readonly type: string;

  /**
   * Used to generate the object.
   * TODO: Instead of BehaviorSubject, this should be split into Observable and Subject/
   */
  readonly payload: {readonly [K in keyof P]: BehaviorSubject<P[K]>};
}
