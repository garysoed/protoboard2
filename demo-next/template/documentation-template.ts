import {cache} from 'gs-tools/export/data';
import {mapNullableTo} from 'gs-tools/export/rxjs';
import {renderTheme} from 'mask';
import {Context, Ctrl, H3, iattr, id, otext, registerCustomElement} from 'persona';
import {Observable} from 'rxjs';

import template from './documentation-template.html';


const $documentationTemplate = {
  host: {
    label: iattr('label'),
  },
  shadow: {
    title: id('title', H3, {
      text: otext(),
    }),
  },
};

class DocumentationTemplate implements Ctrl {
  constructor(private readonly $: Context<typeof $documentationTemplate>) { }

  @cache()
  get runs(): ReadonlyArray<Observable<unknown>> {
    return [
      renderTheme(this.$),
      this.$.host.label.pipe(mapNullableTo(''), this.$.shadow.title.text()),
    ];
  }
}

export const DOCUMENTATION_TEMPLATE = registerCustomElement({
  ctrl: DocumentationTemplate,
  spec: $documentationTemplate,
  tag: 'pbd-documentation-template',
  template,
});
