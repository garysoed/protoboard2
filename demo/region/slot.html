<style>
  #slotContainer {
    display: inline-block;
    position: relative;
    width: 12rem;
    height: 12rem;
  }
</style>

<pbd-documentation-template mk-body-1 label="Slot">
  <div slot="content">
    <p>
      A slot is a container that can contain one piece. Though it is designed to display one piece,
      you can still add multiple pieces, though the item's positions in the slot can be
      unpredictable.
    </p>
    <p>
      <div id="slotContainer">
        <pb-surface id="slot"></pb-surface>
      </div>
    </p>
  </div>
</pbd-documentation-template>