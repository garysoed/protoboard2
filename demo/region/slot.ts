import {$stateService, source} from 'grapevine';
import {StateId} from 'gs-tools/export/state';
import {BaseThemedCtrl, _p} from 'mask';
import {element, PersonaContext} from 'persona';
import {Observable, of} from 'rxjs';

import {$slot, Slot, slotSpec, SlotSpec} from '../../export';
import {DocumentationTemplate} from '../template/documentation-template';

import template from './slot.html';


export const $slotDemo = {
  tag: 'pbd-slot',
  api: {},
};

type State = SlotSpec;

const $state = source<StateId<State>>(
    'state',
    vine => $stateService.get(vine).modify(x => x.add(slotSpec({}, x))),
);

const $ = {
  slot: element('slot', $slot, {}),
};


@_p.customElement({
  ...$slotDemo,
  template,
  dependencies: [
    DocumentationTemplate,
    Slot,
  ],
})
export class SlotDemo extends BaseThemedCtrl<typeof $> {
  constructor(context: PersonaContext) {
    super(context, $);
  }

  protected get renders(): ReadonlyArray<Observable<unknown>> {
    return [
      this.renderers.slot.objectId(of($state.get(this.context.vine))),
    ];
  }
}