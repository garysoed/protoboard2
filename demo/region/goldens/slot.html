<pbd-slot><style>
  #slotContainer {
    display: inline-block;
    position: relative;
    width: 12rem;
    height: 12rem;
  }
</style>

<pbd-documentation-template mk-body-1="" label="Slot"><style>
  #content::slotted(*) {
    padding-left: 1rem;
    padding-right: 1rem;
  }

  #title {
    padding-left: 1rem;
    padding-right: 1rem;
  }
</style>

<section flex="">
  <h3 id="title" mk-display="">Slot</h3>
  <slot id="content" name="content"><div slot="content">
    <p>
      A slot is a container that can contain one piece. Though it is designed to display one piece,
      you can still add multiple pieces, though the item's positions in the slot can be
      unpredictable.
    </p>
    <p>
      </p><div id="slotContainer">
        <pb-surface id="slot" object-path="i::2"><style>
  #root {
    align-items: center;
    justify-content: center;
    background-color: var(--mkThemeBG);
    border: 1px solid var(--mkThemeFG);
    box-sizing: border-box;
    flex-wrap: wrap;
  }
</style>

<div id="root" fill-parent="" mk-action-2="" layout="row">
  <!-- #content-->
</div>
<link rel="stylesheet" href="base/test.css"></pb-surface>
      </div>
    <p></p>
  </div></slot>
</section>
<link rel="stylesheet" href="base/test.css"></pbd-documentation-template><link rel="stylesheet" href="base/test.css"></pbd-slot>